csp.py

```diff
@@ -744,3 +747,526 @@ def solve_zebra(algorithm=min_conflicts, **args):
+                self.var_to_const[var].add(con)
+
+    def __str__(self):
+        """string representation of CSP"""
+        return str(self.domains)
+
+    def display(self, assignment=None):
+        """more detailed string representation of CSP"""
+        if assignment is None:
+            assignment = {}
+        print('CSP(' + str(self.domains) + ', ' + str([str(c) for c in self.constraints]) + ') with assignment: ' +
+              str(assignment))
+
+    def consistent(self, assignment):
+        """assignment is a variable:value dictionary
+        returns True if all of the constraints that can be evaluated
+                        evaluate to True given assignment.
+        """
+        return all(con.holds(assignment)
+                   for con in self.constraints
+                   if all(v in assignment for v in con.scope))
+
+
+class Constraint:
+    """A Constraint consists of
+    * scope: a tuple of variables
+    * condition: a function that can applied to a tuple of values
+    for the variables
+    """
+
+    def __init__(self, scope, condition):
+        self.scope = scope
+        self.condition = condition
+
+    def __repr__(self):
+        return self.condition.__name__ + str(self.scope)
+
+    def holds(self, assignment):
+        """Returns the value of Constraint con evaluated in assignment.
+
+        precondition: all variables are assigned in assignment
+        """
+        return self.condition(*tuple(assignment[v] for v in self.scope))
+
+    @staticmethod
+    def ne_(val):
+        """Returns a function that is True when x is not equal to val, False otherwise"""
+
+        def nev(x):
+            return val != x
+
+        nev.__name__ = str(val) + "!="
+        return nev
+
+    @staticmethod
+    def is_(val):
+        """Returns a function that is True when x is equal to val, False otherwise"""
+
+        def isv(x):
+            return val == x
+
+        isv.__name__ = str(val) + "=="
+        return isv
+
+    @staticmethod
+    def sum_(n):
+        """Returns a function that is True when the the sum of all values is n, False otherwise"""
+
+        def sumv(*values):
+            return sum(values) is n
+
+        sumv.__name__ = str(n) + "==sum"
+        return sumv
+
+    @staticmethod
+    def adjacent(x, y):
+        """Returns True if x and y are adjacent numbers, False otherwise"""
+        return abs(x - y) == 1
+
+    @staticmethod
+    def meet_at(p1, p2):
+        """Returns a function that is True when the words meet at the positions (p1, p2), False otherwise"""
+
+        def meets(w1, w2):
+            return w1[p1] == w2[p2]
+
+        meets.__name__ = "meet_at(" + str(p1) + ',' + str(p2) + ')'
+        return meets
+
+    @staticmethod
+    def is_word(words):
+        """Returns True if the letters concatenated form a word in words, False otherwise"""
+
+        def isw(*letters):
+            return "".join(letters) in words
+
+        return isw
+
+    @staticmethod
```

Comment:
Is it important to mark these as `staticmethod`? I don't think we use it anywhere else in the project and I would prefer to keep things simple.